inmoose/cohort_qc/cohort_metric.py

```diff
@@ -0,0 +1,864 @@
+            The generated plot object.
+        """
+        # Filter the clinical data to select samples corresponding to the current covariate combination
+        filter_condition = (
+            self.clinical_df[self.covariates].eq(list(combination)).all(axis=1)
+        )
+        selected_samples = self.clinical_df[filter_condition].index
+
+        data_after = self.data_expression_df[selected_samples]
+        data_after_melted = data_after.melt(var_name="Sample", value_name="Expression")
+        data_after_melted["Dataset"] = self.clinical_df.loc[
+            data_after_melted["Sample"], self.batch_column
+        ].values
+        data_after_melted["Type"] = data_after_melted["Dataset"].apply(
+            lambda x: "Mixed" if x in self.mixed_datasets else "Non-Mixed"
+        )
+        data_after_melted["Dataset"] = data_after_melted["Dataset"].apply(
+            truncate_name, 11
+        )
+
+        # Plot the boxplot
+        # Calculate the number of unique datasets
+        num_datasets = len(data_after_melted["Dataset"].unique())
+
+        # Dynamically adjust the width of the plot based on the number of datasets
+        base_width = 3  # base width for a small number of datasets
+        width_per_dataset = 0.65  # additional width per dataset
+        fig_width = base_width + width_per_dataset * max(num_datasets - 1, 0)
+
+        if self.data_expression_df_before is not None:
+            fig, axs = plt.subplots(nrows=2, ncols=1, figsize=(fig_width, 8))
+
+            data_before = self.data_expression_df_before[selected_samples]
+            data_before_melted = data_before.melt(
+                var_name="Sample", value_name="Expression"
+            )
+            data_before_melted["Dataset"] = self.clinical_df.loc[
+                data_before_melted["Sample"], self.batch_column
+            ].values
+            data_before_melted["Type"] = data_before_melted["Dataset"].apply(
+                lambda x: "Mixed" if x in self.mixed_datasets else "Non-Mixed"
+            )
+            data_before_melted["Dataset"] = data_before_melted["Dataset"].apply(
+                truncate_name, 11
+            )
+            # Plot before correction
+            sns.violinplot(
+                ax=axs[0],
+                x="Dataset",
+                y="Expression",
+                hue="Type",
+                data=data_before_melted,
+                palette={"Mixed": "red", "Non-Mixed": "grey"},
+            )
+            axs[0].set_title("Gene expression distribution before correction")
+            axs[0].set_xlabel("Dataset (Batch)")
+            axs[0].set_ylabel("Gene Expression")
+            axs[0].tick_params(axis="x", rotation=45)
+            axs[0].legend(title="Dataset Type", loc="upper right")
+            ax = axs[1]
+        else:
+            fig, ax = plt.subplots(nrows=1, ncols=1, figsize=(fig_width, 5))
+
+        # Plot after correction
+        sns.violinplot(
+            ax=ax,
```

Comment:
Does it work when we have also the before dataframe? Since it's called axs and not ax

Reply:
Yes, the code line `ax = axs[1]` when we have to plot before and after allow to use the second part of the subplot